src/libmongoc/src/mongoc/mongoc-topology-description.c

```diff
@@ -649,6 +654,57 @@ _mongoc_topology_description_validate_max_staleness (
    return true;
 }
 
+static bool
+_check_any_server_less_than_wire_version_13 (const void *sd_,
+                                             void *any_too_old_)
+{
+   const mongoc_server_description_t *sd = sd_;
+   bool *any_too_old = any_too_old_;
+   if (sd->max_wire_version < 13) {
+      *any_too_old = true;
+   }
+   return true;
+}
+
+/**
+ * @brief Calculate the read mode that we should be using, based on what was
+ * requested and what is available in the topology.
+ *
+ * Per the CRUD spec, if the requested read mode is *not* primary, and *any*
+ * server in the topology has a wire version <13, we must override the read
+ * mode preference with "primary." Wire version 13 indicates support on a
+ * secondary server for using aggregate pipelines that contain writing stages
+ * (i.e. '$out' and '$merge').
+ */
+static mongoc_read_mode_t
+_calc_effective_read_mode (const mongoc_topology_description_t *td,
+                           mongoc_ss_optype_t optype,
+                           mongoc_read_mode_t requested_read_mode)
+{
+   if (requested_read_mode == MONGOC_READ_PRIMARY ||
+       requested_read_mode == MONGOC_READ_PRIMARY_PREFERRED) {
+      return requested_read_mode;
+   }
+   switch (optype) {
+   case MONGOC_SS_WRITE:
+      return (mongoc_read_mode_t) 0;
```

Comment:
What clause in the [CRUD spec](https://github.com/mongodb/specifications/blob/3a4a6e0f318545c9f050a0000fbd53fc4d9a2790/source/crud/crud.rst#read-preferences-and-server-selection) does this condition ("unset" result) correspond to? My understanding is that this function should either return either `primary` (for pre-5.0 server and etc.) or the "available read preference" (`requested_read_mode`) only. Shouldn't this function be equivalent to the following?
```c
static mongoc_read_mode_t
_calc_effective_read_mode (const mongoc_topology_description_t *td,
                           mongoc_ss_optype_t optype,
                           mongoc_read_mode_t requested_read_mode)
{
   /* This section is only applicable if an explicit (i.e. per-operation) or
    * inherited (e.g. from a Collection) read preference is available and it is
    * not a primary read preference. */
   if (requested_read_mode == MONGOC_READ_PRIMARY) {
      return requested_read_mode;
   }

   /* If there are one or more available servers and one or more of those
    * servers is pre-5.0, drivers MUST NOT use the available read preference and
    * MUST instead select a server using a primary read preference. */
   {
      bool any_too_old = false;
      mongoc_set_for_each_const (mc_tpld_servers_const (td),
                                 _check_any_server_less_than_wire_version_13,
                                 &any_too_old);
      if (any_too_old) {
         return MONGOC_READ_PRIMARY;
      }
   }

   /* Otherwise, if there are either no available servers, all available servers
    * are 5.0+ (i.e. wire version >= 13), or the topology type is LoadBalanced
    * (we can assume the backing mongos is 5.0+), drivers MUST use the available
    * read preference. */
   BSON_ASSERT (mongoc-set-size (mc_tpld_servers_const (td))));
   BSON_ASSERT (!any-too-old); /* Addressed in above block. */
   BSON_ASSERT (is-loadbalanced (td)); /* Per LoadBalanced assumption. */
   return requested_read_mode;
}
```

It is also not clear to me from the spec what the behavior should be if the topology type is _not_ LoadBalanced (which seems like a possibility that is not ruled out by prior conditions). Possible spec wording defect?

Reply:
This is another good question. The purpose of calculating the effective read mode here is to send that information back to the caller so that it can modify its behavior when it builds the command to send to the server. Importantly, the overriding of the read mode should only apply if the intended operation is an aggregate operation with write stages, so that branch _must_ remain (A regular old read operation should stick to a secondary if the user requested a secondary). 

Whether to care about the read mode when the intended operation is a write (as with `MONGOC_SS_WRITE`) is interesting, since it should have no bearing on server selection. Using the zero "unset" value when the operation is a write is just working under the assumption that the user won't be giving use a readPreference for a write operation, since that doesn't seem to make sense.